File: weknot_android/app/src/main/java/com/example/weknot_android/network/comm/SocialComm.java
package com.example.weknot_android.network.comm;

import com.example.weknot_android.base.BaseComm;
import com.example.weknot_android.network.api.SocialApi;
import com.example.weknot_android.room.entity.user.Friend;

import java.util.List;

import io.reactivex.Single;

public class SocialComm extends BaseComm<SocialApi> {

    public Single<List<Friend>> getFriends(String token) {
        return api.getFriends(token).map(getResponseObjectsFunction());
    }

    @Override
    protected Class<SocialApi> type() {
        return SocialApi.class;
    }
}

File: weknot_android/app/src/main/java/com/example/weknot_android/network/comm/SocialComm.kt
package com.example.weknot_android.network.comm

import com.example.weknot_android.base.BaseComm
import com.example.weknot_android.model.chat.PrivateChatRoom
import com.example.weknot_android.model.user.Friend
import com.example.weknot_android.network.api.SocialApi
import com.example.weknot_android.network.request.FriendAddRequest
import com.example.weknot_android.network.request.FriendRequest
import com.google.firebase.database.FirebaseDatabase
import io.reactivex.Single

class SocialComm : BaseComm<SocialApi>() {

    fun getFriends(token: String): Single<List<Friend>> {
        return api.getFriends(token).map(getResponseObjectsFunction())
    }

    fun putFriend(token: String, userId: String, friendRequest: FriendRequest): Single<String> {

        val chatRoomUid = FirebaseDatabase.getInstance().reference.child("privatechatrooms").push().key!!
        FirebaseDatabase.getInstance().reference.child("privatechatrooms").child(chatRoomUid).setValue(PrivateChatRoom(userId, friendRequest.friend))

        return api.putFriend(token, friendRequest).map(getResponseMessageFunction())
    }

    fun addFriend(token: String, receiver: String): Single<String> {
        return api.addFriend(token, FriendAddRequest(receiver)).map(getResponseMessageFunction())
    }

    override fun type(): Class<SocialApi> {
        return SocialApi::class.java
    }
}